src/meta/src/hummock/compaction/picker/min_overlap_compaction_picker.rs

```diff
@@ -163,210 +167,203 @@ impl NonOverlapSubLevelPicker {
-                    break;
-                }
-
-                last_level_index = target_index;
-                overlap_len_and_begins.push(overlap_files_range);
-
-                if overlap_files.is_empty() {
-                    // We allow a layer in the middle without overlap, so we need to continue to
-                    // the next layer to search for overlap
-                    continue;
-                }
-
-                level_select_files[target_index].extend(overlap_files.into_iter());
-                select_level_count += 1;
+                overlap_info.update(other);
+                select_sst_id_set.insert(other.sst_id);
+                current_level_size += other.file_size;
             }
 
             if pending_compact {
                 break;
             }
 
-            overlap_len_and_begins.pop();
+            let mut extra_overlap_levels = vec![];
 
+            let mut add_files_size = 0;
             // check reverse overlap
-            for (reverse_index, old_overlap_range) in (0..last_level_index)
-                .rev()
-                .zip_eq_fast(overlap_len_and_begins.into_iter().rev())
+            for (reverse_index, old_overlap_range) in
+                overlap_len_and_begins.iter_mut().enumerate().rev()
             {
-                let target_tables = &l0[reverse_index].table_infos;
+                let target_tables = &levels[reverse_index].table_infos;
+                // It has select all files in this sub-level, so it can not overlap with more files.
+                if ret.sstable_infos[reverse_index].len() == target_tables.len() {
+                    continue;
+                }
                 let new_overlap_range = overlap_info.check_multiple_overlap(target_tables);
                 let mut extra_overlap_sst = Vec::with_capacity(new_overlap_range.len());
-                for new_overlap_index in new_overlap_range {
+                for new_overlap_index in new_overlap_range.clone() {
                     if old_overlap_range.contains(&new_overlap_index) {
                         // Since some of the files have already been selected when selecting
                         // upwards, we filter here to avoid adding sst repeatedly
                         continue;
                     }
 
                     let other = &target_tables[new_overlap_index];
-
                     if level_handler.is_pending_compact(&other.sst_id) {
                         pending_compact = true;
                         break;
                     }
-
                     debug_assert!(!select_sst_id_set.contains(&other.sst_id));
-
-                    all_file_size += other.file_size;
-                    all_file_count += 1;
-
+                    add_files_size += other.file_size;
                     overlap_info.update(other);
                     select_sst_id_set.insert(other.sst_id);
                     extra_overlap_sst.push(other.clone());
                 }
 
```

Comment:
need to check `pending_compact`, we no need to check next reverse_level if pending_compact